x-pack/filebeat/input/httpjson/requester.go

```diff
@@ -0,0 +1,464 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package httpjson
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+
+	cursor "github.com/elastic/beats/v7/filebeat/input/v2/input-cursor"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type requestInfo struct {
+	url        string
+	contentMap common.MapStr
+	headers    common.MapStr
+}
+
+type retryLogger struct {
+	log *logp.Logger
+}
+
+func newRetryLogger() *retryLogger {
+	return &retryLogger{
+		log: logp.NewLogger("httpjson.retryablehttp", zap.AddCallerSkip(1)),
+	}
+}
+
+func (l *retryLogger) Printf(s string, args ...interface{}) {
+	l.log.Debugf(s, args...)
+}
+
+type requester struct {
+	log         *logp.Logger
+	config      config
+	client      *http.Client
+	cursorValue string
+}
```

Comment:
this type does a lot. It implements rate limiting, request execution, response parsing+event creation, and publishing. It looks like it is very difficult to test. Can we split some functionality into a dedicated type and introduce unit testing? At least for rate limiting consider to use/introduce a third party library.

Reply:
For the rate limit bit, it was not in the scope of this change, but it makes sense, will open a new issue to make it separately if that makes sense.